libctru/source/system/syscalls.c

```diff
@@ -28,6 +31,64 @@ struct _reent* __SYSCALL(getreent)()
 	return tv->reent;
 }
 
+int __SYSCALL(clock_gettime)(clockid_t clock_id, struct timespec *tp) {
+
+	if (clock_id == CLOCK_REALTIME)
+	{
+		if (tp != NULL)
+		{
+			// Retrieve current time, adjusting epoch from 1900 to 1970
+			s64 ms_since_epoch = osGetTime() - 2208988800000ULL;
+			tp->tv_sec = ms_since_epoch / 1000;
+			tp->tv_nsec = (ms_since_epoch - 1000*tp->tv_sec) * 1000000;
+		}
+	}
+	else if (clock_id == CLOCK_MONOTONIC) 
+	{
+		if (tp != NULL)
+		{
+			// Use the ticks directly, as it offer the highest precision
+			u64 usec_since_boot = svcGetSystemTick() / CPU_TICKS_PER_USEC;
+
+			tp->tv_sec = usec_since_boot / 1000000;
+			tp->tv_nsec = (usec_since_boot - 1000000*tp->tv_sec) * 1000;
```

Comment:
My version:
```cpp
            u64 ticks_since_boot = svcGetSystemTick();

            tp->tv_sec  = ticks_since_boot / SYSCLOCK_ARM11;
            tp->tv_nsec = (ticks_since_boot % SYSCLOCK_ARM11) * 1000000000ULL / SYSCLOCK_ARM11;
```
This gives about 3.7ns resolution.

I don't know why, but your version does some floating-point conversions but mine didn't.

Reply:
Thanks, that's much better!